### PR TITLE
[CtrlPkt] Convert `control_packet` to `half_dma_cpy_nd` operations

### PR DESCRIPTION
**Pull request by Yu-Zhewen**
From control_packet operations, we extract two types of information:
- **reconfiguration sequence**: includes the header and data necessary for reconfiguration.
- **DMA operations:** covers memcpy and synchronization to send the content into the device through the shim.

The DMA operations can be further lowered to transactions using existing controlcode passes.

References: https://github.com/Xilinx/mlir-aie/pull/1709; https://github.com/Xilinx/mlir-aie/pull/1753

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+      SmallVector<int64_t> offsets{0, 0, 0,
+                                   static_cast<long>(ctrlPktSequence.size())};
+      SmallVector<int64_t> sizes{1, 1, 1, headerAndDataLength};
+      SmallVector<int64_t> strides{0, 0, 0, 1};
+
+      // Store the control packet header and data.
+      llvm::MutableArrayRef<uint32_t> words =
+          reserveAndGetTail(headerAndDataLength);
+      // Subtract 1 from `dataLength` because the length `i` is encoded in the
+      // header as `i - 1`.
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp:109]: You could hide the `- 1` subtraction inside the API?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+struct ControlPacketDmaBuilder {
+  AMDAIE::AMDAIEDeviceModel deviceModel;
+  ControlPacketDmaBuilder(AMDAIE::AMDAIEDeviceModel deviceModel)
+      : deviceModel(std::move(deviceModel)) {}
+
+  std::vector<uint32_t> ctrlPktSequence;
+
+  llvm::MutableArrayRef<uint32_t> reserveAndGetTail(size_t tailSize) {
+    auto oldSize = ctrlPktSequence.size();
+    auto newSize = oldSize + tailSize;
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp:30]:
```suggestion
    size_t oldSize = ctrlPktSequence.size();
    size_t newSize = oldSize + tailSize;
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitControlPacketData.cpp [hunk 3]
```diff
@@ -87,7 +87,10 @@ void AMDAIESplitControlPacketDataPass::runOnOperation() {
   }
   AMDAIE::AMDAIEDeviceModel deviceModel =
       AMDAIE::getDeviceModel(maybeDevice.value());
-  uint32_t maxLength = deviceModel.getCtrlPktMaxLength();
+  // Add 1 to the `beat` field value, as it represents `length - 1` in the
+  // control packet header. For example, if the maximum `beat` field value is
+  // 15, the corresponding maximum length is 16.
+  uint32_t maxLength = deviceModel.getCtrlPktMaxBeat() + 1;
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitControlPacketData.cpp:93]: I think `getCtrlPktMaxLength` is more descriptive, so could we keep it like that and do the `+ 1` addition inside that function?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      }
+    }
+    return
+  }
+}
+
+// -----
+
+// CHECK:       0x80032000
+// CHECK:       0x00000000
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir:36]: I think it would be useful to add some test cases with multiple and larger control packets as well

File: runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc [hunk 5]
```diff
@@ -316,9 +316,48 @@ uint32_t AMDAIEDeviceModel::getCoreTileLocalMemorySize() const {
   return devInst.DevProp.DevMod[XAIEGBL_TILE_TYPE_AIETILE].CoreMod->DataMemSize;
 }
 
-uint32_t AMDAIEDeviceModel::getCtrlPktMaxLength() const {
-  return 2 << (ctrlPktHeaderFormat.operationShift -
-               ctrlPktHeaderFormat.beatShift);
+uint32_t AMDAIEDeviceModel::getCtrlPktHeader(uint32_t address, uint32_t beat,
```
Comment (jtuyls) [inline runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc:319]: I think returning an error would be better than asserting here as it would be possible for a user function to provide incorrect arguments (address too large) in a release build. Also, the checks could change based on the device generations etc. So, you could return `FailureOr<uint32_t>` instead.

**Review by jtuyls — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    res = workgroupOp->walk([&](AMDAIE::NpuControlPacketOp ctrlPktOp) {
+      ctrlPktOps.push_back(ctrlPktOp);
+      // Get `ConnectionOp` for the `CTRL` port.
+      uint32_t colShift = deviceModel.getColumnShift();
+      uint32_t rowShift = deviceModel.getRowShift();
+      llvm::errs() << "colShift: " << colShift << "rowShift: " << rowShift
+                   << "\n";
+      uint32_t address = ctrlPktOp.getAddress() & 0xFFFFF;
+      int32_t col = (ctrlPktOp.getAddress() >> colShift) & 0x1F;
+      int32_t row = (ctrlPktOp.getAddress() >> rowShift) & 0x1F;
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp]: implemented here
https://github.com/nod-ai/iree-amd-aie/blob/ad577e8cc6d5f6f824f732258492ca6162d0e428/runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc#L543-L556

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp (hunk 1 above)
Comment (Yu-Zhewen) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp]: Thanks, done.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp (hunk 2 above)
Comment (Yu-Zhewen) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlPacketToHalfDmaCpyNd.cpp]: done

File: runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc (hunk 5 above)
Comment (Yu-Zhewen) [reply runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc]: Done

https://github.com/nod-ai/iree-amd-aie/blob/ad577e8cc6d5f6f824f732258492ca6162d0e428/runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc#L319

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir (hunk 4 above)
Comment (Yu-Zhewen) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir:36]: Add one test for checking the `length out of range` situation:
https://github.com/nod-ai/iree-amd-aie/blob/ad577e8cc6d5f6f824f732258492ca6162d0e428/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir#L80

and another more complex test with more control packets:
https://github.com/nod-ai/iree-amd-aie/blob/ad577e8cc6d5f6f824f732258492ca6162d0e428/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/control_packet_to_half_dma_cpy_nd.mlir#L106

**Review by jtuyls — APPROVED** [review]
LGTM, thanks!
